### ISSUE TITLE
`filter().collect()` returns empty when `scan_csv()` reads large file

### ISSUE THREAD

**Issue by Mr-Milk** [pola-rs/polars#1808]
#### Are you using Python or Rust?

**Python**

#### What version of polars are you using?

v0.10.21

#### What operating system are you using polars on?

Windows 11

#### Describe your bug.

Lazy API. `filter().collect()` returns empty when `scan_csv()` reads large file

The polars works well with small lazy dataframe

<img width="301" alt="image" src="https://user-images.githubusercontent.com/23433306/142359390-dd136774-b6a5-4e8f-aa55-d98bedf9f5bd.png">

When I read a local csv file that's ~18G, a similar operation returns empty. 

To make sure that the record exists, I perform a unique operation that returns the result.

<img width="425" alt="image" src="https://user-images.githubusercontent.com/23433306/142359548-6d29210a-d544-41c5-80aa-6f64e85668b4.png">

I also tried `dask` to perform the task

<img width="429" alt="image" src="https://user-images.githubusercontent.com/23433306/142359757-d03cafc7-c93f-4d35-89c0-a06c8bffd4fc.png">

Sorry about posting pictures here, I don't know how to share a large file.

**Comment by ritchie46** [comment]
Could you try recreating tgis problem on a subset of your file and then share that?

**Comment by Mr-Milk** [comment]
That's the problem, if I use a subset of the file, `polars` works perfectly. I can't reproduce the behavior on a small file.

**Comment by ritchie46** [comment]
can you somehow compress and share it?

**Comment by ghuls** [comment]
@ritchie46 I think it might be the same bug has in https://github.com/pola-rs/polars/issues/1777 where something goes wrong during parsing of the CSV file in columns.

```python
In [10]: pl.read_csv('atac_fragments.tsv.gz', sep='\t', has_headers=False, skip_rows=100).filter(pl.col("column_4") == "GTGAACAAGACAGGCG-1")
Out[10]: 
shape: (18811, 5)
┌────────────┬──────────┬──────────┬────────────────────┬──────────┐
│ column_1   ┆ column_2 ┆ column_3 ┆ column_4           ┆ column_5 │
│ ---        ┆ ---      ┆ ---      ┆ ---                ┆ ---      │
│ str        ┆ i64      ┆ i64      ┆ str                ┆ i64      │
╞════════════╪══════════╪══════════╪════════════════════╪══════════╡
│ chr1       ┆ 3123846  ┆ 3123985  ┆ GTGAACAAGACAGGCG-1 ┆ 2        │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ chr1       ┆ 3301956  ┆ 3302127  ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ chr1       ┆ 3400999  ┆ 3401138  ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ chr1       ┆ 3485540  ┆ 3485723  ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ ...        ┆ ...      ┆ ...      ┆ ...                ┆ ...      │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ JH584304.1 ┆ 59601    ┆ 59656    ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ JH584304.1 ┆ 59725    ┆ 59821    ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ JH584304.1 ┆ 59866    ┆ 60019    ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ GL456216.1 ┆ 16207    ┆ 16510    ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ GL456216.1 ┆ 32011    ┆ 32046    ┆ GTGAACAAGACAGGCG-1 ┆ 2        │
└────────────┴──────────┴──────────┴────────────────────┴──────────┘

In [11]: pl.scan_csv('atac_fragments.tsv.gz', sep='\t', has_headers=False, skip_rows=100).filter(pl.col("column_4") == "GTGAACAAGACAGGCG-1").collect()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-11-0534ead65075> in <module>
----> 1 pl.scan_csv('atac_fragments.tsv.gz', sep='\t', has_headers=False, skip_rows=100).filter(pl.col("column_4") == "GTGAACAAGACAGGCG-1").collect()

/staging/leuven/stg_00002/lcb/ghuls/software/miniconda3/envs/create_cistarget_databases/lib/python3.8/site-packages/polars/lazy/frame.py in collect(self, type_coercion, predicate_pushdown, projection_pushdown, simplify_expression, string_cache, no_optimization)
    297             string_cache,
    298         )
--> 299         return pl.eager.frame.wrap_df(ldf.collect())
    300 
    301     def fetch(

RuntimeError: Any(ComputeError("invalid utf8 data in csv"))
```

**Comment by ritchie46** [comment]
@ghuls, Thanks I will take a look at that first.

**Comment by ghuls** [comment]
@ritchie46 Seems pl.scan_csv does not support compression of the input file.

With input file gunzipped, it now parsers it without visible errors, but returns way less rows than it should:

```
In [20]: pl.read_csv('atac_fragments.tsv', sep='\t', has_headers=False, skip_rows=100).filter(pl.col("column_4") == "GTGAACAAGACAGGCG-1")
Out[20]: 
shape: (18811, 5)
┌────────────┬──────────┬──────────┬────────────────────┬──────────┐
│ column_1   ┆ column_2 ┆ column_3 ┆ column_4           ┆ column_5 │
│ ---        ┆ ---      ┆ ---      ┆ ---                ┆ ---      │
│ str        ┆ i64      ┆ i64      ┆ str                ┆ i64      │
╞════════════╪══════════╪══════════╪════════════════════╪══════════╡
│ chr1       ┆ 3123846  ┆ 3123985  ┆ GTGAACAAGACAGGCG-1 ┆ 2        │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ chr1       ┆ 3301956  ┆ 3302127  ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ chr1       ┆ 3400999  ┆ 3401138  ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ chr1       ┆ 3485540  ┆ 3485723  ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ ...        ┆ ...      ┆ ...      ┆ ...                ┆ ...      │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ JH584304.1 ┆ 59601    ┆ 59656    ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ JH584304.1 ┆ 59725    ┆ 59821    ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ JH584304.1 ┆ 59866    ┆ 60019    ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ GL456216.1 ┆ 16207    ┆ 16510    ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ GL456216.1 ┆ 32011    ┆ 32046    ┆ GTGAACAAGACAGGCG-1 ┆ 2        │
└────────────┴──────────┴──────────┴────────────────────┴──────────┘

In [21]: pl.scan_csv('atac_fragments.tsv', sep='\t', has_headers=False, skip_rows=100).filter(pl.col("column_4") == "GTGAACAAGACAGGCG-1").collect()
Out[21]: 
shape: (18, 5)
┌──────────┬───────────┬───────────┬────────────────────┬──────────┐
│ column_1 ┆ column_2  ┆ column_3  ┆ column_4           ┆ column_5 │
│ ---      ┆ ---       ┆ ---       ┆ ---                ┆ ---      │
│ str      ┆ i64       ┆ i64       ┆ str                ┆ i64      │
╞══════════╪═══════════╪═══════════╪════════════════════╪══════════╡
│ chr1     ┆ 3123846   ┆ 3123985   ┆ GTGAACAAGACAGGCG-1 ┆ 2        │
├╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ chr1     ┆ 3301956   ┆ 3302127   ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ chr1     ┆ 3400999   ┆ 3401138   ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ chr1     ┆ 3485540   ┆ 3485723   ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ ...      ┆ ...       ┆ ...       ┆ ...                ┆ ...      │
├╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ chr3     ┆ 135838847 ┆ 135838906 ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ chr5     ┆ 133023332 ┆ 133023491 ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ chr8     ┆ 4476999   ┆ 4477150   ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ chr8     ┆ 4586515   ┆ 4586644   ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
├╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
│ chr8     ┆ 4656761   ┆ 4656861   ┆ GTGAACAAGACAGGCG-1 ┆ 1        │
└──────────┴───────────┴───────────┴────────────────────┴──────────┘
```

**Comment by ritchie46** [comment]
> @ritchie46 Seems pl.scan_csv does not support compression of the input file.

Ah, yes! We must throw an error in such a case. Compression is all or nothing. :/

**Comment by Mr-Milk** [comment]
I use a smaller file and compress it. It's 33MB in `.gz`, here is the [link](https://uofmacau-my.sharepoint.com/:u:/g/personal/yb97643_umac_mo/EZjE2aN8N-FOvYrHsv_cEvwBIyLLT5myv1ZtnyoMce75gg?e=uiBMv5)

**Comment by tempbottle** [comment]
I had a problem like this when I read a large csv file via lazy api. After do some fix like this:
```rust
let lf = LazyCsvReader::new(file_path.into())
        .has_header(true)
        .with_delimiter(b',')
        .finish();
 let lf_open = lf.clone().collect()?.lazy().filter(col("is_open").gt(lit(0i64)));
 let df_open = lf_open.collect()?;
```
It works fine.

Note `collect()?.lazy()` in line ` let lf_open = lf.clone().collect()?.lazy().filter(col("is_open").gt(lit(0i64)));` is needed here.

Otherwise, I got a df_open with `smaller` height than DataFrame version:
```rust
let lf = LazyCsvReader::new(file_path.into())
        .has_header(true)
        .with_delimiter(b',')
        .finish();
let df = lf.clone().collect()?;
let dfmask = df.column("is_open")?.gt(0i32);
let df_open = df.filter(&dfmask)?;
```

Something like before lazy Expr execute, we must load full csv file at least once here.

**Comment by Mr-Milk** [comment]
> I had a problem like this when I read a large csv file via lazy api. After do some fix like this:
> 
> ```rust
> let lf = LazyCsvReader::new(file_path.into())
>         .has_header(true)
>         .with_delimiter(b',')
>         .finish();
>  let lf_open = lf.clone().collect()?.lazy().filter(col("is_open").gt(lit(0i64)));
>  let df_open = lf_open.collect()?;
> ```
> 
> It works fine.
> 
> Note `collect()?.lazy()` in line ` let lf_open = lf.clone().collect()?.lazy().filter(col("is_open").gt(lit(0i64)));` is needed here.
 
> Otherwise, I got a df_open with `smaller` height than DataFrame version:
> 
> ```rust
> let lf = LazyCsvReader::new(file_path.into())
>         .has_header(true)
>         .with_delimiter(b',')
>         .finish();
> let df = lf.clone().collect()?;
> let dfmask = df.column("is_open")?.gt(0i32);
> let df_open = df.filter(&dfmask)?;
> ```
> 
> Something like before lazy Expr execute, we must load full csv file at least once here.


If we need to collect it and turn it to lazy again, then it's no different than using the eager API.

**Comment by tempbottle** [comment]
> > I had a problem like this when I read a large csv file via lazy api. After do some fix like this:
> > ```rust
> > let lf = LazyCsvReader::new(file_path.into())
> >         .has_header(true)
> >         .with_delimiter(b',')
> >         .finish();
> >  let lf_open = lf.clone().collect()?.lazy().filter(col("is_open").gt(lit(0i64)));
> >  let df_open = lf_open.collect()?;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It works fine.
> > Note `collect()?.lazy()` in line ` let lf_open = lf.clone().collect()?.lazy().filter(col("is_open").gt(lit(0i64)));` is needed here.
> 
> > Otherwise, I got a df_open with `smaller` height than DataFrame version:
> > ```rust
> > let lf = LazyCsvReader::new(file_path.into())
> >         .has_header(true)
> >         .with_delimiter(b',')
> >         .finish();
> > let df = lf.clone().collect()?;
> > let dfmask = df.column("is_open")?.gt(0i32);
> > let df_open = df.filter(&dfmask)?;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Something like before lazy Expr execute, we must load full csv file at least once here.
> 
> If we need to collect it and turn it to lazy again, then it's no different than using the eager API.

Right, this is not lazy style.

**Comment by ritchie46** [comment]
@Mr-Milk I have a fix in #1817. After this is merged I will immediately release a patch.

**Comment by Mr-Milk** [comment]
> @Mr-Milk I have a fix in #1817. After this is merged I will immediately release a patch.

Thanks, That's quick!
